### ISSUE TITLE
Maintenance task time keeps getting bigger

### ISSUE THREAD

**Issue by mjandjel** [timescale/promscale#1282]
Dear community, we are noticing that the Promscale maintenance jobs are taking too long and that the times reported in postgres logs are increasing:

```
$ grep "promscale maintenance jobs"  /var/log/postgres/postgres.log
Mar 11 20:50:37 lxromdbv3p postgres[1640885]: [9-1] 2022-03-11 20:50:37.041 CET [1640885] WARNING:  promscale maintenance jobs are taking too long (one run took 12:17:13.295491)
Mar 24 06:21:25 lxpromdbv3p postgres[3934278]: [9-1] 2022-03-24 06:21:25.266 CET [3934278] WARNING:  promscale maintenance jobs are taking too long (one run took 13:49:01.564494)
Mar 25 08:34:33 lxpromdbv3p postgres[4071280]: [12-1] 2022-03-25 08:34:33.463 CET [4071280] WARNING:  promscale maintenance jobs are taking too long (one run took 18:52:49.699599)
Apr  2 11:31:42 lxpromdbv3p postgres[1236515]: [8-1] 2022-04-02 11:31:42.157 CEST [1236515] WARNING:  promscale maintenance jobs are taking too long (one run took 14:18:08.976219)
Apr  2 11:34:45 lxpromdbv3p postgres[1236504]: [8-1] 2022-04-02 11:34:45.069 CEST [1236504] WARNING:  promscale maintenance jobs are taking too long (one run took 14:21:18.917581)
Apr  3 00:17:38 lxpromdbv3p postgres[1378886]: [8-1] 2022-04-03 00:17:38.993 CEST [1378886] WARNING:  promscale maintenance jobs are taking too long (one run took 12:15:56.686601)
Apr  3 00:20:08 lxpromdbv3p postgres[1379220]: [8-1] 2022-04-03 00:20:08.746 CEST [1379220] WARNING:  promscale maintenance jobs are taking too long (one run took 12:15:23.647307)
Apr  3 17:30:31 lxpromdbv3p postgres[1465949]: [8-1] 2022-04-03 17:30:31.494 CEST [1465949] WARNING:  promscale maintenance jobs are taking too long (one run took 16:42:52.445432)
Apr  3 17:42:19 lxpromdbv3p postgres[1466241]: [8-1] 2022-04-03 17:42:19.665 CEST [1466241] WARNING:  promscale maintenance jobs are taking too long (one run took 16:52:10.883653)
Apr  4 10:32:15 lxpromdbv3p postgres[1580116]: [8-1] 2022-04-04 10:32:15.499 CEST [1580116] WARNING:  promscale maintenance jobs are taking too long (one run took 16:19:55.783282)
Apr  4 10:35:04 lxpromdbv3p postgres[1578795]: [8-1] 2022-04-04 10:35:04.451 CEST [1578795] WARNING:  promscale maintenance jobs are taking too long (one run took 16:34:32.872454)
Apr  5 06:55:07 lxpromdbv3p postgres[1692562]: [8-1] 2022-04-05 06:55:07.596 CEST [1692562] WARNING:  promscale maintenance jobs are taking too long (one run took 19:50:03.032229)
Apr  5 07:18:24 lxpromdbv3p postgres[1692255]: [8-1] 2022-04-05 07:18:24.467 CEST [1692255] WARNING:  promscale maintenance jobs are taking too long (one run took 20:16:08.899475)

```

Our infrastructure is already well described in issue #1214, but we put basic extension versions here: 

```
 oid  |   extname   | extowner | extnamespace | extrelocatable | extversion |                                                                                 extconfig
                                               |                            extcondition
-------+-------------+----------+--------------+----------------+------------+----------------------------------------------------------------------------------------------------------------------------
-----------------------------------------------+--------------------------------------------------------------------
 13420 | plpgsql     |       10 |           11 | f              | 1.0        |
                                               |
 20063 | promscale   |       10 |        18588 | f              | 0.3.0      |
                                               |
 16387 | timescaledb |       10 |         2200 | f              | 2.6.0      | {16404,16406,16427,16441,16456,16454,16475,16473,16491,16489,16513,16529,16531,16547,16559,16561,16604,16612,16633,16646,16
656,16660,16677,16696,16711,16734,16740,16737} | {"","","","","","","","","","","","","","","","WHERE id >= 1000","+
       |             |          |              |                |            |
                                               |   WHERE KEY = 'exported_uuid' ","","","","","","","","","","",""}
(3 rows)

```
 

The configuration for the maintenance tasks is a default one:

```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | total_r
uns | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+--------
----+-----------------+----------------
                   |                 |      1 | 2022-04-05 06:07:30.370675+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.207109   | 2022-04-06 02:54:26.077784+02 |
 52 |               0 |             52
                   |                 |   1004 | 2022-04-05 07:48:24.476317+02 | 2022-04-05 07:18:24.473679+02 |                 | Scheduled  |                   | -infinity                     |
 29 |              26 |              0
                   |                 |   1005 | 2022-04-05 07:25:07.700151+02 | 2022-04-05 06:55:07.693177+02 |                 | Scheduled  |                   | -infinity                     |
 29 |              26 |              0
(3 rows)



timescale=# SELECT * FROM timescaledb_information.jobs;
 job_id |      application_name      | schedule_interval | max_runtime | max_retries | retry_period |      proc_schema      |        proc_name        |  owner   | scheduled | config |          next_star
t           | hypertable_schema | hypertable_name
--------+----------------------------+-------------------+-------------+-------------+--------------+-----------------------+-------------------------+----------+-----------+--------+-------------------
------------+-------------------+-----------------
      1 | Telemetry Reporter [1]     | 24:00:00          | 00:01:40    |          -1 | 01:00:00     | _timescaledb_internal | policy_telemetry        | postgres | t         |        | 2022-04-06 02:54:2
6.077784+02 |                   |
   1004 | User-Defined Action [1004] | 00:30:00          | 00:00:00    |          -1 | 00:05:00     | _prom_catalog         | execute_maintenance_job | postgres | t         |        | -infinity
            |                   |
   1005 | User-Defined Action [1005] | 00:30:00          | 00:00:00    |          -1 | 00:05:00     | _prom_catalog         | execute_maintenance_job | postgres | t         |        | -infinity
            |                   |
(3 rows)

timescale=#
```

The timescale database size:
```timescale=# SELECT pg_size_pretty( pg_database_size('timescale') );
 pg_size_pretty
----------------
 560 GB
(1 row)
```

Number of hypertables and chunks:
```
timescale=# select count(*) from  _timescaledb_catalog.hypertable;
 count
-------
  8316
(1 row)

timescale=# select count(*) from  _timescaledb_catalog.chunk;
 count
--------
 670631
(1 row)

timescale=#
```
Our database host is having 32 CPU cores/192GB or RAM. Two Promscale instances are ingesting around 100K samples persecond in average.

Can you please recommend which maintenance interval size should we use and what number of maintenance tasks should we configure in our environment to improve the performance of maintenance jobs?

**Comment by antekresic** [comment]
Hi there.

I think you should definitely increase the number of jobs here. Jobs are most likely falling behind behind the chunk creation expecially if they are taking over 12 hours while default chunk interval is at 8 hours. 

Since you have 32 CPU, you might want to increase them to 8 and see what that does. I don't know the load you have on the database, 8 seems like a good starting point. After that point, I'd suggest trying to gauge if that is not enough or maybe too much. 

You can do that by using `config_maintenance_jobs` function like so:
`select config_maintenance_jobs(8, interval '30 min');`
That should gracefully restart any running jobs and start them up.

Since we are doing this, we should also double check that you have increased the `max_worker_processes` PG setting. Default is set at 8, I would suggest upping that as well to at least 12. More details can be found here: https://www.postgresql.org/docs/current/runtime-config-resource.html#GUC-MAX-WORKER-PROCESSES

Again, these are all recommendations which are based on the info you've sent us. Fine tuning them would depend on the actual load you are seeing after the recommendations are applied.

Hope this helps, do let us know about the outcome here.

Thanks!

**Comment by mjandjel** [comment]
Hi @antekresic , thank you for the recommendations! Today we tried incrementing the number of maintenance jobs from 2 to 8. 

```
timescale=# select config_maintenance_jobs(8, interval '30 min');
 config_maintenance_jobs
-------------------------
 t
(1 row)


timescale=#  SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | total_r
uns | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+--------
----+-----------------+----------------
                   |                 |      1 | 2022-04-06 02:54:26.142536+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.2517     | 2022-04-06 21:20:44.394236+02 |
 53 |               0 |             53
                   |                 |   1004 | 2022-04-05 18:04:30.654292+02 | 2022-04-05 07:18:24.473679+02 |                 | Scheduled  |                   | -infinity                     |
 30 |              26 |              0
                   |                 |   1005 | 2022-04-06 07:04:03.918645+02 | 2022-04-06 06:34:03.910881+02 |                 | Scheduled  |                   | -infinity                     |
 30 |              27 |              0
(3 rows)

timescale=#  SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | total_r
uns | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+--------
----+-----------------+----------------
                   |                 |      1 | 2022-04-06 02:54:26.142536+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.2517     | 2022-04-06 21:20:44.394236+02 |
 53 |               0 |             53
                   |                 |   1008 | 2022-04-06 09:21:27.558467+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  1 |               0 |              0
                   |                 |   1009 | 2022-04-06 09:21:27.569706+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  1 |               0 |              0
                   |                 |   1004 | 2022-04-05 18:04:30.654292+02 | 2022-04-05 07:18:24.473679+02 |                 | Scheduled  |                   | -infinity                     |
 30 |              26 |              0
                   |                 |   1005 | 2022-04-06 07:04:03.918645+02 | 2022-04-06 06:34:03.910881+02 |                 | Scheduled  |                   | -infinity                     |
 30 |              27 |              0
                   |                 |   1010 | 2022-04-06 09:21:27.581445+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  1 |               0 |              0
                   |                 |   1011 | 2022-04-06 09:21:27.456768+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  1 |               0 |              0
(7 rows)
```

First we tried this on local test database where immediately 8 workers appeared as a result. On production database we see only seven job records after running the statement.
Our machine has 32 cpu which is top limit for the machine. We did not notice significant increase in CPU load after the tasks were started. We'll need a few days for system to catch up with the maintenance and will report if the maintenance job run time has lowered. 

We have checked Postgres config as a possible clue for not getting 8 workers as requested but our parallel worker numbers do not appear too small:
```
$  patronictl -c /etc/patroni.yml show-config
loop_wait: 10
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    autovacuum_max_workers: 10
    autovacuum_naptime: 10
    checkpoint_completion_target: 0.9
    default_statistics_target: 500
    effective_cache_size: 75GB
    effective_io_concurrency: 200
    maintenance_work_mem: 2047MB
    max_connections: 10000
    max_locks_per_transaction: 512
    max_parallel_workers: 16
    max_parallel_workers_per_gather: 8
    max_wal_size: 100GB
    max_worker_processes: 27
    min_wal_size: 512MB
    random_page_cost: 1.1
    shared_buffers: 25GB
    synchronous_commit: false
    timescaledb.max_background_workers: 8
    track_activities: false
    track_counts: true
    wal_buffers: 16MB
    wal_compression: true
    wal_keep_size: 800GB
    work_mem: 64MB
  timescaledb.telemetry_level: false
  use_pg_rewind: true
  use_slots: true
retry_timeout: 10
ttl: 30
```

Any comments on what could be preventing us from reaching 8 maintenance jobs?

**Comment by mjandjel** [comment]
We restarted the databases and now have 8 maintenance jobs running:

```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | total_r
uns | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+--------
----+-----------------+----------------
                   |                 |      1 | 2022-04-06 02:54:26.142536+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.2517     | 2022-04-06 21:20:44.394236+02 |
 53 |               0 |             53
                   |                 |   1006 | 2022-04-06 11:23:24.520333+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  1 |               0 |              0
                   |                 |   1007 | 2022-04-06 11:23:24.527094+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  1 |               0 |              0
                   |                 |   1008 | 2022-04-06 11:28:59.678449+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  2 |               0 |              0
                   |                 |   1009 | 2022-04-06 11:28:29.211091+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  2 |               0 |              0
                   |                 |   1004 | 2022-04-06 11:34:16.081981+02 | 2022-04-05 07:18:24.473679+02 |                 | Scheduled  |                   | -infinity                     |
 31 |              26 |              0
                   |                 |   1005 | 2022-04-06 11:28:24.522575+02 | 2022-04-06 06:34:03.910881+02 |                 | Scheduled  |                   | -infinity                     |
 31 |              27 |              0
                   |                 |   1010 | 2022-04-06 11:28:24.524409+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  2 |               0 |              0
                   |                 |   1011 | 2022-04-06 11:28:40.925681+02 | -infinity                     |                 | Scheduled  |                   | -infinity                     |
  2 |               0 |              0
(9 rows)


```

**Comment by mjandjel** [comment]
After running for around two days with 8 maintenance jobs we notice that the time taken by the jobs is **around 10 hours** which is better comparing to old setup with two jobs and 20 hours per run.

We checked job_stats table periodically to track job execution:

```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration
|          next_start           | total_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------
+-------------------------------+------------+-----------------+----------------
                   |                 |      1 | 2022-04-07 06:27:34.401888+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.142687
| 2022-04-08 01:31:22.544575+02 |         54 |               0 |             54
                   |                 |   1006 | 2022-04-07 07:30:00.089546+02 | 2022-04-07 07:00:00.082158+02 |                 | Scheduled  |
| -infinity                     |          3 |               2 |              0
                   |                 |   1007 | 2022-04-07 06:57:33.436719+02 | 2022-04-07 06:27:33.433596+02 |                 | Scheduled  |
| -infinity                     |          3 |               2 |              0
                   |                 |   1008 | 2022-04-07 07:02:28.903676+02 | 2022-04-07 06:32:28.899031+02 |                 | Scheduled  |
| -infinity                     |          4 |               2 |              0
                   |                 |   1009 | 2022-04-07 07:15:21.262073+02 | 2022-04-07 06:45:21.255374+02 |                 | Scheduled  |
| -infinity                     |          4 |               2 |              0
                   |                 |   1004 | 2022-04-07 07:49:53.609169+02 | 2022-04-07 07:19:53.605085+02 |                 | Scheduled  |
| -infinity                     |         33 |              28 |              0
                   |                 |   1005 | 2022-04-07 07:03:58.725877+02 | 2022-04-07 06:33:58.719767+02 |                 | Scheduled  |
| -infinity                     |         33 |              29 |              0
                   |                 |   1010 | 2022-04-07 07:04:11.449484+02 | 2022-04-07 06:34:11.445417+02 |                 | Scheduled  |
| -infinity                     |          4 |               2 |              0
                   |                 |   1011 | 2022-04-07 07:07:37.011648+02 | 2022-04-07 06:37:37.009072+02 |                 | Scheduled  |
| -infinity                     |          4 |               2 |              0
(9 rows)

```

```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration
|          next_start           | total_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------
+-------------------------------+------------+-----------------+----------------
                   |                 |      1 | 2022-04-07 06:27:34.401888+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.142687
| 2022-04-08 01:31:22.544575+02 |         54 |               0 |             54
                   |                 |   1006 | 2022-04-07 17:09:39.200801+02 | 2022-04-07 16:39:39.194106+02 |                 | Scheduled  |
| -infinity                     |          4 |               3 |              0
                   |                 |   1007 | 2022-04-07 17:23:02.596001+02 | 2022-04-07 16:53:02.59167+02  |                 | Scheduled  |
| -infinity                     |          4 |               3 |              0
                   |                 |   1008 | 2022-04-07 17:10:41.100642+02 | 2022-04-07 16:40:41.096257+02 |                 | Scheduled  |
| -infinity                     |          5 |               3 |              0
                   |                 |   1009 | 2022-04-07 17:18:12.262775+02 | 2022-04-07 16:48:12.26044+02  |                 | Scheduled  |
| -infinity                     |          5 |               3 |              0
                   |                 |   1004 | 2022-04-07 17:15:33.201061+02 | 2022-04-07 16:45:33.19711+02  |                 | Scheduled  |
| -infinity                     |         34 |              29 |              0
                   |                 |   1005 | 2022-04-07 18:03:10.843157+02 | 2022-04-07 17:33:10.840178+02 |                 | Scheduled  |
| -infinity                     |         34 |              30 |              0
                   |                 |   1010 | 2022-04-07 17:50:39.211559+02 | 2022-04-07 17:20:39.204047+02 |                 | Scheduled  |
| -infinity                     |          5 |               3 |              0
                   |                 |   1011 | 2022-04-07 17:17:27.09267+02  | 2022-04-07 16:47:27.087471+02 |                 | Scheduled  |
| -infinity                     |          5 |               3 |              0
(9 rows)
```



```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | tot
al_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+----
--------+-----------------+----------------
                   |                 |      1 | 2022-04-08 02:56:09.740415+02 | -infinity                     | Failed          | Scheduled  | 00:00:03.13222    | 2022-04-08 21:59:57.872635+02 |
     55 |               0 |             55
                   |                 |   1006 | 2022-04-08 03:32:05.576211+02 | 2022-04-08 03:02:05.570518+02 |                 | Scheduled  |                   | -infinity                     |
      5 |               4 |              0
                   |                 |   1007 | 2022-04-08 03:33:30.326844+02 | 2022-04-08 03:03:30.319805+02 |                 | Scheduled  |                   | -infinity                     |
      5 |               4 |              0
                   |                 |   1008 | 2022-04-08 03:37:33.667888+02 | 2022-04-08 03:07:33.662911+02 |                 | Scheduled  |                   | -infinity                     |
      6 |               4 |              0
                   |                 |   1009 | 2022-04-08 03:42:18.787056+02 | 2022-04-08 03:12:18.780542+02 |                 | Scheduled  |                   | -infinity                     |
      6 |               4 |              0
                   |                 |   1004 | 2022-04-08 03:34:11.292826+02 | 2022-04-08 03:04:11.290045+02 |                 | Scheduled  |                   | -infinity                     |
     35 |              30 |              0
                   |                 |   1005 | 2022-04-08 03:26:08.83149+02  | 2022-04-08 02:56:08.826087+02 |                 | Scheduled  |                   | -infinity                     |
     35 |              31 |              0
                   |                 |   1010 | 2022-04-08 03:28:17.111505+02 | 2022-04-08 02:58:17.10722+02  |                 | Scheduled  |                   | -infinity                     |
      6 |               4 |              0
                   |                 |   1011 | 2022-04-08 03:46:18.957633+02 | 2022-04-08 03:16:18.950683+02 |                 | Scheduled  |                   | -infinity                     |
      6 |               4 |              0
(9 rows)

```

Since we are still much behind the desired performance of maintenance job, we are not sure if increasing them further would bring additional improvement. One side effect of this setup appears to be increased CPU iowait times on DB server that correlate with times when jobs are started. During these spikes, duplicates metric on Promscale spikes also and Prometheuses clusters switch leaders. 

Following image shows DB server CPU iowait spikes correlating with job start times:
![image](https://user-images.githubusercontent.com/23310076/162381488-e41cf49f-fcc0-4db3-aaad-26a789c58b37.png)

If this spikes are indeed caused by many maintenance jobs doing massive data queries at their start, we would need to disperse job start times to try eliminating iowait spikes.  @antekresic any recommendations on how to disperse starts of these maintenance jobs?

**Comment by mjandjel** [comment]
Update:
Thanks to Promscale team advice, we managed to start all 8 maintenance jobs separated by 15 minutes.

```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      | last_successful_finish | last_run_status | job_status | last_run_duration |          next_start           | total_runs
 | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+------------------------+-----------------+------------+-------------------+-------------------------------+-----------
-+-----------------+----------------
                   |                 |      1 | 2022-04-08 02:56:09.740415+02 | -infinity              | Failed          | Scheduled  | 00:00:03.13222    | 2022-04-08 21:59:57.872635+02 |         55
 |               0 |             55
                   |                 |   1012 | 2022-04-08 13:50:51.023552+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          2
 |               0 |              0
                   |                 |   1013 | 2022-04-08 13:50:53.376684+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          2
 |               0 |              0
                   |                 |   1014 | 2022-04-08 12:46:35.571994+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
                   |                 |   1015 | 2022-04-08 13:50:53.874187+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          2
 |               0 |              0
                   |                 |   1016 | 2022-04-08 13:51:07.939937+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          2
 |               0 |              0
                   |                 |   1017 | 2022-04-08 13:30:32.726582+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
                   |                 |   1018 | 2022-04-08 13:45:55.989029+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
                   |                 |   1019 | 2022-04-08 14:00:41.671886+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
(9 rows)

```

This was achieved by running following statements around 15 minutes apart:

**Comment by mjandjel** [comment]
Few moments later:

**Comment by mjandjel** [comment]
Error in jobs have caused start times to get closer for the errors-affected jobs. We won't be repeating the whole startup process again. We'll leave it to run like so and monitor system behavior.

**Comment by mjandjel** [comment]
Update:

After receiving a performance hotfix from Promscale, we installed it on old production database and let it run from Friday to Sunday. We started 3 maintenance jobs:
```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start
   | total_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+----------------------------
---+------------+-----------------+----------------
                   |                 |      1 | 2022-03-18 14:26:11.984135+01 | -infinity                     | Failed          | Scheduled  | 00:00:00.022181   | 2022-03-19 10:07:27.006316+
01 |        255 |               0 |            255
                   |                 |   1075 | 2022-04-10 10:51:07.042312+02 | 2022-04-10 11:04:06.511391+02 | Success         | Scheduled  | 00:12:59.469079   | 2022-04-10 11:34:06.511391+
02 |         79 |              79 |              0
                   |                 |   1076 | 2022-04-10 10:51:03.797478+02 | 2022-04-10 11:03:00.215698+02 | Success         | Scheduled  | 00:11:56.41822    | 2022-04-10 11:33:00.215698+
02 |         79 |              79 |              0
                   |                 |   1077 | 2022-04-10 10:51:03.794748+02 | 2022-04-10 11:03:00.238194+02 | Success         | Scheduled  | 00:11:56.443446   | 2022-04-10 11:33:00.238194+
02 |         78 |              78 |              0
(4 rows)
```
As it can be seen from the output above, processes worked flawlessly, 79/79 times with last run duration of around 11 minutes. The database received full production ingestion load. Avg CPU load was around 40% which is excellent.

Since we wanted to test how would the system behave if we remove the patch, we applied patch to return us to the vanilla 0.10 Promscale version at around 11:00 AM, same day. Again, 3 maintenance jobs were configured to run.
```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      | last_successful_finish | last_run_status | job_status | last_run_duration |          next_start           | to
tal_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+------------------------+-----------------+------------+-------------------+-------------------------------+---
---------+-----------------+----------------
                   |                 |      1 | 2022-03-18 14:26:11.984135+01 | -infinity              | Failed          | Scheduled  | 00:00:00.022181   | 2022-03-19 10:07:27.006316+01 |
     255 |               0 |            255
                   |                 |   1078 | 2022-04-10 11:09:47.74473+02  | -infinity              |                 | Scheduled  |                   | -infinity                     |
       1 |               0 |              0
                   |                 |   1079 | 2022-04-10 11:11:22.949279+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |
       1 |               0 |              0
(3 rows)

```

Hoping to collect some useful job stats, the system was checked again @ around 18:30:
```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start
   | total_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+----------------------------
---+------------+-----------------+----------------
                   |                 |      1 | 2022-03-18 14:26:11.984135+01 | -infinity                     | Failed          | Scheduled  | 00:00:00.022181   | 2022-03-19 10:07:27.006316+
01 |        255 |               0 |            255
                   |                 |   1078 | 2022-04-10 18:11:59.206801+02 | 2022-04-10 17:41:59.202146+02 |                 | Scheduled  |                   | -infinity
   |          2 |               1 |              0
                   |                 |   1079 | 2022-04-10 18:19:09.365257+02 | 2022-04-10 17:49:09.356597+02 |                 | Scheduled  |                   | -infinity
   |          2 |               1 |              0
                   |                 |   1080 | 2022-04-10 13:06:37.424048+02 | -infinity                     |                 | Scheduled  |                   | -infinity
   |          1 |               0 |              0
(4 rows)

Without any traces of errors in Postgres logs, all attempts to run the processes ended up in failure for unknown cause and we could not obtain results on how the old code runs but the same code on the production database needs around 10h to complete the run and this is accompanied by spikes in CPU I/O waits and DB access share locks.



```

**Comment by mjandjel** [comment]
We again applied the new hotfix and hour later checked for the results:
```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      | last_successful_finish | last_run_status | job_status | last_run_duration |          next_start           | to
tal_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+------------------------+-----------------+------------+-------------------+-------------------------------+---
---------+-----------------+----------------
                   |                 |      1 | 2022-03-18 14:26:11.984135+01 | -infinity              | Failed          | Scheduled  | 00:00:00.022181   | 2022-03-19 10:07:27.006316+01 |
     255 |               0 |            255
                   |                 |   1081 | 2022-04-10 19:09:30.303843+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |
       1 |               0 |              0
                   |                 |   1082 | 2022-04-10 19:15:20.27869+02  | -infinity              |                 | Scheduled  |                   | -infinity                     |
       1 |               0 |              0
                   |                 |   1083 | 2022-04-10 19:16:34.05891+02  | -infinity              |                 | Scheduled  |                   | -infinity                     |
       1 |               0 |              0
(4 rows)

```
Again, new code working flawlessly with short durations:
```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start
   | total_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+----------------------------
---+------------+-----------------+----------------
                   |                 |      1 | 2022-03-18 14:26:11.984135+01 | -infinity                     | Failed          | Scheduled  | 00:00:00.022181   | 2022-03-19 10:07:27.006316+
01 |        255 |               0 |            255
                   |                 |   1081 | 2022-04-10 20:01:19.756885+02 | 2022-04-10 20:04:50.20676+02  | Success         | Scheduled  | 00:03:30.449875   | 2022-04-10 20:34:50.20676+0
2  |          2 |               2 |              0
                   |                 |   1082 | 2022-04-10 20:02:39.192768+02 | 2022-04-10 20:04:52.713475+02 | Success         | Scheduled  | 00:02:13.520707   | 2022-04-10 20:34:52.713475+
02 |          2 |               2 |              0
                   |                 |   1083 | 2022-04-10 20:01:21.24106+02  | 2022-04-10 20:04:44.322727+02 | Success         | Scheduled  | 00:03:23.081667   | 2022-04-10 20:34:44.322727+
02 |          2 |               2 |              0
(4 rows)

```

**Comment by mjandjel** [comment]
Update: we applied patch on production database starting 4 maintenance jobs:

```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      | last_successful_finish | last_run_status | job_status | last_run_duration |          next_start           | total_runs
 | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+------------------------+-----------------+------------+-------------------+-------------------------------+-----------
-+-----------------+----------------
                   |                 |      1 | 2022-04-10 19:05:07.28815+02  | -infinity              | Failed          | Scheduled  | 00:00:03.120526   | 2022-04-11 16:48:17.908676+02 |         58
 |               0 |             58
                   |                 |   1020 | 2022-04-11 09:16:40.408806+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
                   |                 |   1021 | 2022-04-11 09:17:45.482845+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
                   |                 |   1022 | 2022-04-11 09:22:45.797556+02 | -infinity              |                 | Scheduled  |                   | -infinity                     |          1
 |               0 |              0
(4 rows)

```


Unlike with the test database, here the maintenance jobs have also to process metric retentions:
```
timescale=# select application_name || '(' || backend_type || ')' as who, count(*) as cnt from pg_stat_activity where datname='timescale' group by who;
                                             who                                             | cnt
---------------------------------------------------------------------------------------------+-----
 (autovacuum worker)                                                                         |  10
 promscale@0.10.0(client backend)                                                            |  25
 promscale maintenance: data retention: metric mysql_global_stat(User-Defined Action [1022]) |   1
 promscale maintenance: data retention: metric mysql_global_vari(User-Defined Action [1020]) |   1
 promscale maintenance: data retention: metric mysql_global_vari(User-Defined Action [1021]) |   1
 promscale maintenance: data retention(User-Defined Action [1023])                           |   1
 psql(client backend)                                                                        |   1
 TimescaleDB Background Worker Scheduler(TimescaleDB Background Worker Scheduler)            |   1
(8 rows)

timescale=#
```

**Comment by mjandjel** [comment]
Seems like some kind of locking occurs during maintenance where one long standing job processing huge node_cpu metric is preventing other jobs to continue:
```
timescale=# select application_name || '(' || backend_type || ')' as who, count(*) as cnt from pg_stat_activity where datname='timescale' group by who;
                                             who                                             | cnt
---------------------------------------------------------------------------------------------+-----
 (autovacuum worker)                                                                         |  10
 (client backend)                                                                            |   1
 promscale@0.10.0(client backend)                                                            |  31
 promscale maintenance: compression: metric node_cpu(User-Defined Action [1022])             |   1
 promscale maintenance: compression: metric node_cpu: waiting on(User-Defined Action [1020]) |   1
 promscale maintenance: compression: metric node_cpu: waiting on(User-Defined Action [1021]) |   1
 promscale maintenance: compression: metric node_cpu: waiting on(User-Defined Action [1023]) |   1
 psql(client backend)                                                                        |   1
 TimescaleDB Background Worker Scheduler(TimescaleDB Background Worker Scheduler)            |   1

```

But, few minutes later:

**Comment by mjandjel** [comment]
timescale=# SELECT pg_size_pretty( pg_database_size('timescale') );
 pg_size_pretty
----------------
 455 GB
(1 row)


timescale=# SELECT pg_size_pretty( pg_database_size('timescale') );
 pg_size_pretty
----------------
 409 GB
(1 row)

**Comment by mjandjel** [comment]
Second cycle of maintenance has completed with even smaller times:
```
timescale=# SELECT * FROM timescaledb_information.job_stats;
 hypertable_schema | hypertable_name | job_id |      last_run_started_at      |    last_successful_finish     | last_run_status | job_status | last_run_duration |          next_start           | tot
al_runs | total_successes | total_failures
-------------------+-----------------+--------+-------------------------------+-------------------------------+-----------------+------------+-------------------+-------------------------------+----
--------+-----------------+----------------
                   |                 |      1 | 2022-04-10 19:05:07.28815+02  | -infinity                     | Failed          | Scheduled  | 00:00:03.120526   | 2022-04-11 16:48:17.908676+02 |
     58 |               0 |             58
                   |                 |   1020 | 2022-04-11 10:29:14.548655+02 | 2022-04-11 10:42:50.150343+02 | Success         | Scheduled  | 00:13:35.601688   | 2022-04-11 11:12:50.150343+02 |
      2 |               2 |              0
                   |                 |   1021 | 2022-04-11 10:29:14.730151+02 | 2022-04-11 10:42:44.336135+02 | Success         | Scheduled  | 00:13:29.605984   | 2022-04-11 11:12:44.336135+02 |
      2 |               2 |              0
                   |                 |   1022 | 2022-04-11 10:32:45.209995+02 | 2022-04-11 10:42:44.339666+02 | Success         | Scheduled  | 00:09:59.129671   | 2022-04-11 11:12:44.339666+02 |
      2 |               2 |              0
                   |                 |   1023 | 2022-04-11 10:29:14.514994+02 | 2022-04-11 10:42:44.333591+02 | Success         | Scheduled  | 00:13:29.818597   | 2022-04-11 11:12:44.333591+02 |
      2 |               2 |              0
(5 rows)

timescale=# SELECT pg_size_pretty( pg_database_size('timescale') );
 pg_size_pretty
----------------
 407 GB
(1 row)
```

**Comment by mjandjel** [comment]
Now the CPU usage is lower, slowdowns of ingestion/duplicates no longer appear.
Promscale dashboard for past 24 hours shows how the performance of the system is improving after the patch took effect after 9AM:
![image](https://user-images.githubusercontent.com/23310076/162700612-3b07337c-e00d-44a5-889c-906173d75525.png)

**Comment by mwolter805** [comment]
Hello, I’ve noticed the same performance issue related to promscale maintenance jobs when using version 0.10.0. Also noticed this approximately six months ago when testing and decided to simply have Prometheus store the data. My use case is injecting Prometheus metrics from my home assistant home automation system, which should be somewhat simple for promscale to handle. 

My needs have grown somewhat beyond promql and that’s why I’m trying pro scale again. Anyway, at the time of maintenance, CPU usage jumps from 3% to approximately 25% and appears to increase over time as the promscale database grows. Not a huge issue but it will grow to become one. 

This hotfix sounds promising, is it possible to apply this hotfix to a Docker installation?

**Comment by ramonguiu** [comment]
@mwolter805 yes, the hotfix can be applied to Docker installations. We will be publishing instructions on how to apply it today.

**Comment by ramonguiu** [comment]
@mwolter805 details to apply the hotfix can be found [here](https://github.com/timescale/promscale/blob/master/docs/compression-job-hotfix.md)

Please open a new issue if it doesn't work for you

**Comment by ramonguiu** [comment]
I'm closing this issue now
